### ISSUE TITLE
ComputeResidual.cpp:60:5: error: variable 'n' must have explicitly specified data sharing attributes

### ISSUE THREAD

**Issue by jdayal-ssi** [hpcg-benchmark/hpcg#81]
I am trying to compile hpcg 3.1 on an Ampere neoverse_n1. I am trying to build it with clang, mpich and openmp.
Versions: 

mpich 4.1.1
clang: 16.0.6
OS: rocky linux 8

CXXFLAGS: -O3 -ffast-math -ftree-vectorize -fopenmp

When I try to compile, I get the following errors and warnings:

jai.dayal@sal-amperehn01:~/hpcg-3.1/neoverse_n1$ make
mpicxx -c  -I./src -I./src/neoverse_n1  -O3 -ffast-math -ftree-vectorize -fopenmp -I../src ../src/ComputeResidual.cpp -o src/ComputeResidual.o
../src/ComputeResidual.cpp:60:5: error: variable 'n' must have explicitly specified data sharing attributes
    for (local_int_t i=0; i<n; i++) {
    ^~~
../src/ComputeResidual.cpp:56:32: note: explicit data sharing attribute requested here
  #pragma omp parallel default(none) shared(local_residual, v1v, v2v)
                               ^
1 error generated.
make: *** [Makefile:76: src/ComputeResidual.o] Error 1

Any advice? Thanks in advance.

**Comment by jdayal-ssi** [comment]
Fixed this by changing the line 
shared(local_residual, v1v, v2v,)
to
shared(local_residual, v1v, v2v, n)

**Comment by maherou** [comment]
The variable 'n' should be shared.  Try adding it to the list of shared variables, just after v2v.
________________________________
From: jdayal-ssi ***@***.***>
Sent: Friday, September 22, 2023 5:09 PM
To: hpcg-benchmark/hpcg ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [EXTERNAL] [hpcg-benchmark/hpcg] ComputeResidual.cpp:60:5: error: variable 'n' must have explicitly specified data sharing attributes (Issue #81)


I am trying to compile hpcg 3.1 on an Ampere neoverse_n1. I am trying to build it with clang, mpich and openmp.
Versions:

mpich 4.1.1
clang: 16.0.6
OS: rocky linux 8

CXXFLAGS: -O3 -ffast-math -ftree-vectorize -fopenmp

When I try to compile, I get the following errors and warnings:

***@***.***:~/hpcg-3.1/neoverse_n1$ make
mpicxx -c -I./src -I./src/neoverse_n1 -O3 -ffast-math -ftree-vectorize -fopenmp -I../src ../src/ComputeResidual.cpp -o src/ComputeResidual.o
../src/ComputeResidual.cpp:60:5: error: variable 'n' must have explicitly specified data sharing attributes
for (local_int_t i=0; i<n; i++) {
^~~
../src/ComputeResidual.cpp:56:32: note: explicit data sharing attribute requested here
#pragma omp parallel default(none) shared(local_residual, v1v, v2v)
^
1 error generated.
make: *** [Makefile:76: src/ComputeResidual.o] Error 1

Any advice? Thanks in advance.

**Comment by maherou** [comment]
Sorry.  Just saw this.  Glad the fix worked.
________________________________
From: jdayal-ssi ***@***.***>
Sent: Friday, September 22, 2023 6:20 PM
To: hpcg-benchmark/hpcg ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [EXTERNAL] Re: [hpcg-benchmark/hpcg] ComputeResidual.cpp:60:5: error: variable 'n' must have explicitly specified data sharing attributes (Issue #81)


Closed #81<https://github.com/hpcg-benchmark/hpcg/issues/81> as completed.

**Comment by luszczek** [comment]
I cannot reproduce this problem with Clang 14 through 16 nor with godbolt.org where I used this source code:
```C++
#include <cmath>

typedef int local_int_t;
struct Vector_STRUCT{local_int_t localLength;double * values;void*optimizationData;};
typedef struct Vector_STRUCT Vector;
int ComputeResidual(const local_int_t n, const Vector & v1, const Vector & v2, double & residual) {

  double * v1v = v1.values;
  double * v2v = v2.values;
  double local_residual = 0.0;

#pragma omp parallel shared(local_residual, v1v, v2v)
  {
    double threadlocal_residual = 0.0;
    #pragma omp for
    for (local_int_t i=0; i<n; i++) {
      double diff = std::fabs(v1v[i] - v2v[i]);
      if (diff > threadlocal_residual) threadlocal_residual = diff;
    }
    #pragma omp critical
    {
      if (threadlocal_residual>local_residual) local_residual = threadlocal_residual;
    }
  }
  residual = local_residual;
  return 0;
}
```
It would be good to know the origins of the tool chain used for MPICH 4.1.1. Was it system default installation or it came from another packag manager?
